### PR TITLE
T13060 filter service

### PR DESCRIPTION
**Pull request by niden**
Hello!

* Type: new feature
* Link to issue: #13060 

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)?
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

Rewrote the `Filter` component. The old `Filter` component has been deprecated. The new component uses the `ServiceLocator` and `ServiceLocatorFactory` classes to offer a lightweight component for sanitizing content. Each sanitizer is a separate class, all of them lazy loaded to offer maxumum performance. The component accepts new sanitizers wither when it is instantiated or after it is instantiated. Each sanitizer can be called directly using the magic methods (the name of the sanitizer is the name of the mehtod) or the locator can call the all to familiar `sanitize` method to sanitize content (even arrays) with one or multiple sanitizers

Thanks

### CODE REVIEW

**Comment by niden** [comment]
cc @sergeyklay

File: phalcon/cli/dispatcher.zep [hunk 1]
```diff
@@ -177,7 +178,8 @@ class Dispatcher extends CliDispatcher implements DispatcherInterface
 				CliDispatcher::EXCEPTION_NO_DI
 			);
 		}
-		let filter = <FilterInterface> dependencyInjector->getShared("filter");
+		let filter = <LocatorInterface> dependencyInjector->getShared("filter");
+//		let filter = <FilterInterface> dependencyInjector->getShared("filter");
```
Comment (sergeyklay) [inline phalcon/cli/dispatcher.zep]: Do we need this comment in the code base?

File: phalcon/di/factorydefault.zep [hunk 2]
```diff
@@ -37,7 +41,8 @@ class FactoryDefault extends \Phalcon\Di
 			"eventsManager":      new Service("Phalcon\\Events\\Manager", true),
 			"flash":              new Service("Phalcon\\Flash\\Direct", true),
 			"flashSession":       new Service("Phalcon\\Flash\\Session", true),
-			"filter":             new Service("Phalcon\\Filter", true),
+			"filter":             new Service(filter->newInstance(), true),
+//			"filter":             new Service("Phalcon\\Filter", true),
```
Comment (sergeyklay) [inline phalcon/di/factorydefault.zep]: Do we need this comment in the code base?

File: phalcon/dispatcher.zep [hunk 3]
```diff
@@ -266,7 +267,8 @@ abstract class Dispatcher implements DispatcherInterface, InjectionAwareInterfac
 		if typeof dependencyInjector != "object" {
 			this->{"_throwDispatchException"}("A dependency injection object is required to access the 'filter' service", self::EXCEPTION_NO_DI);
 		}
-		let filter = <FilterInterface> dependencyInjector->getShared("filter");
+		let filter = <LocatorInterface> dependencyInjector->getShared("filter");
+//		let filter = <FilterInterface> dependencyInjector->getShared("filter");
```
Comment (sergeyklay) [inline phalcon/dispatcher.zep]: Do we need this comment in the code base?

File: tests/unit/Http/Request/GetPostCest.php [hunk 4]
```diff
@@ -30,6 +33,23 @@ class GetPostCest
...
+
+        $container = new FactoryDefault();
+        Di::reset();
+        Di::setDefault($container);
+
+        $request = $container['request'];
+
+//        var_dump($container['filter']); die();
+
+//        $request = $this->getRequestObject();
```
Comment (sergeyklay) [inline tests/unit/Http/Request/GetPostCest.php]: What is it

File: tests/unit/Http/Request/GetPostCest.php [hunk 5]
```diff
@@ -12,12 +12,15 @@
...
+use Phalcon\Di;
+use Phalcon\Di\FactoryDefault;
+use Phalcon\Test\Unit\Http\Helper\HttpBase;
 use UnitTester;
 
 /**
  * Class GetPostCest
  */
-class GetPostCest
+class GetPostCest //extends HttpBase
```
Comment (sergeyklay) [inline tests/unit/Http/Request/GetPostCest.php]: What is it

File: phalcon/cli/dispatcher.zep (hunk 1 above)
Comment (sergeyklay) [inline phalcon/cli/dispatcher.zep]:
```suggestion
```

File: phalcon/di/factorydefault.zep (hunk 2 above)
Comment (sergeyklay) [inline phalcon/di/factorydefault.zep]:
```suggestion
```

File: phalcon/dispatcher.zep (hunk 3 above)
Comment (sergeyklay) [inline phalcon/dispatcher.zep]:
```suggestion
```

File: phalcon/forms/form.zep [hunk 6]
```diff
@@ -199,7 +200,8 @@ class Form extends Injectable implements \Countable, \Iterator
 
 				if typeof filter != "object" {
 					let dependencyInjector = this->getDI(),
-						filter = <FilterInterface> dependencyInjector->getShared("filter");
+						filter = <LocatorInterface> dependencyInjector->getShared("filter");
+//						filter = <FilterInterface> dependencyInjector->getShared("filter");
```
Comment (sergeyklay) [inline phalcon/forms/form.zep]:
```suggestion
```

File: phalcon/http/cookie.zep [hunk 7]
```diff
@@ -214,7 +215,8 @@ class Cookie implements CookieInterface, InjectionAwareInterface
 							}
 						}
 
-						let filter = dependencyInjector->getShared("filter"),
+//						let filter = dependencyInjector->getShared("filter"),
```
Comment (sergeyklay) [inline phalcon/http/cookie.zep]:
```suggestion
```

File: phalcon/http/request.zep [hunk 8]
```diff
@@ -1073,15 +1074,15 @@ class Request implements RequestInterface, InjectionAwareInterface
...
 		if filters !== null {
 			let filter = this->_filter;
 			if typeof filter != "object" {
 				let dependencyInjector = <DiInterface> this->_dependencyInjector;
 				if typeof dependencyInjector != "object" {
 					throw new Exception("A dependency injection object is required to access the 'filter' service");
 				}
-				let filter = <FilterInterface> dependencyInjector->getShared("filter");
+				let filter = <LocatorInterface> dependencyInjector->getShared("filter");
+//				let filter = <FilterInterface> dependencyInjector->getShared("filter");
```
Comment (sergeyklay) [inline phalcon/http/request.zep]:
```suggestion
```

File: phalcon/validation.zep [hunk 9]
```diff
@@ -520,7 +521,8 @@ class Validation extends Injectable implements ValidationInterface
 					}
 				}
 
-				let filterService = dependencyInjector->getShared("filter");
+				let filterService = <LocatorInterface> dependencyInjector->getShared("filter");
+//				let filterService = dependencyInjector->getShared("filter");
```
Comment (sergeyklay) [inline phalcon/validation.zep]:
```suggestion
```

File: tests/unit/Di/FactoryDefault/ConstructCest.php [hunk 10]
```diff
@@ -51,7 +51,8 @@ private function getServices(): array
             'eventsManager'      => 'Phalcon\Events\Manager',
             'flash'              => 'Phalcon\Flash\Direct',
             'flashSession'       => 'Phalcon\Flash\Session',
-            'filter'             => 'Phalcon\Filter',
+            'filter'             => 'Phalcon\Filter\FilterLocator',
+//            'filter'             => 'Phalcon\Filter',
```
Comment (sergeyklay) [inline tests/unit/Di/FactoryDefault/ConstructCest.php]:
```suggestion
```

File: tests/unit/Http/Request/GetPostCest.php (hunk 5 above)
Comment (sergeyklay) [inline tests/unit/Http/Request/GetPostCest.php]:
```suggestion
class GetPostCest
```

File: tests/unit/Http/Request/GetPostCest.php [hunk 11]
```diff
@@ -30,6 +33,23 @@ class GetPostCest
...
+        $oldPost = $_POST;
+        $_POST['test'] = -1234;
+
+        $container = new FactoryDefault();
+        Di::reset();
+        Di::setDefault($container);
+
+        $request = $container['request'];
+
+//        var_dump($container['filter']); die();
```
Comment (sergeyklay) [inline tests/unit/Http/Request/GetPostCest.php]:
```suggestion
```

File: tests/unit/Http/Request/GetPostCest.php (hunk 4 above)
Comment (sergeyklay) [inline tests/unit/Http/Request/GetPostCest.php]:
```suggestion
```

**Review by sergeyklay — DISMISSED** [review]
Looks good to me (except some commented code)

**Comment by niden** [comment]
Talked to @sergeyklay . We will leave the commented code there just for reference for now and remove it later on.

**Pull request merged**